Document license file expiration logic (#19485)
gravitational/teleport: docs/pages/deploy-a-cluster/teleport-enterprise/license.mdx

---
title: Enterprise License File
description: Teleport enterprise license file configuration parameters and requirements
description: How to manage your Teleport Enterprise license file.
---

> This guide explains the use of a license file with Teleport Enterprise.

Reviewer: Should we add a warning at the top when `scope==cloud` to say that cloud licenses are managed for you and this guide does not apply to cloud?

---
title: Enterprise License File
description: How to manage your Teleport Enterprise license file.
---

Self-hosted Teleport Enterprise subscriptions require a valid license. In this
guide, we will show you how to manage your Teleport Enterprise license file.

(!docs/pages/includes/cloud/call-to-action.mdx!)

## Managing your license file

In a self-hosted Teleport Enterprise cluster, the Teleport Auth Service reads a
license file in order to determine the scope and validity of your subscription.
Follow the steps below to give your Teleport Auth Service instances access to
your license file.

### Download your license file

To obtain your license file, visit the [Teleport customer

> dashboard](https://dashboard.gravitational.com/web/login) and log in. Click

Reviewer 2: @russjones do we have a name for the customer portal?

Author: Looks like "customer portal" is the most common usage, so I'll change it to that:

```
% grep -RohE "\[[A-Za-z ]+\]\(https://dashboard.gravitational.com/web/login\)" | sort | uniq -c | sort -n
   1 [Dashboard](https://dashboard.gravitational.com/web/login)
   1 [customer dashboard](https://dashboard.gravitational.com/web/login)
   3 [Customer Portal](https://dashboard.gravitational.com/web/login)
  10 [customer portal](https://dashboard.gravitational.com/web/login)
```

default.

</TabItem>
<TabItem label="Kubernetes">

Rename your license file to `license.pem`.

Create a secret called "license" in the namespace you are using to deploy
Teleport:

```code
$ kubectl -n teleport create secret generic license --from-file=license.pem
```

For the Teleport pod to load your license:

- The secret must be named `license`
- The secret must be in the same Kubernetes namespace as the Teleport pod
- The license file in the secret to be named `license.pem`

</TabItem>
</Tabs>

Only Auth Service instances require the license. Instances of other Teleport
services do not need a license file unless they are also running the Auth
Service.

### Check your license expiration date

Your Teleport Enterprise license contains an X.509 certificate that contains the
expiration date of your license. 

Run the following command to inspect the expiration date of your license,
assuming your license is saved as `license.pem`:

```code
$ openssl x509 -text -in license.pem | grep "Not After"
             Not After : Dec 16 19:43:40 2022 GMT
```

### Renew or update your license

If you have changed your license agreement with Teleport, e.g., you have added

> or removed support for a feature, you must obtain your license file again and

Reviewer: ```suggestion
or removed support for a feature, you must obtain an update license file and
```

new license. Inaction may lead to unplanned outage or degraded performance and
support.
```

Users will see a similar message as a banner when accessing the Teleport Web UI:

![License warning banner](../../../img/enterprise/license-warning.png)

When your license expires, you will have a 30-day grace period before Teleport
Enterprise features become disabled. 

`tsh` users will see warnings indicating that Enterprise features are no longer
available:

```text
Your Teleport Enterprise Edition license has expired on one or more of your auth
servers. Please reach out to licenses@goteleport.com to obtain a new license.
Inaction may lead to unplanned outage or degraded performance and support.
```

Teleport Web UI users will see an alert banner:

![License expired banner](../../../img/enterprise/license-expired.png)

Attempts to authenticate to Teleport via Single Sign-On connectors will fail.

## Unlicensed Teleport features

If users attempt to use a Teleport feature that your license does not allow,
they will see an error message. For example, if you attempt to use Database
Access without it being included in the license, you will get a message similar
to:

```text

> this Teleport cluster is not licensed for database access, please contact the cluster administrator

Reviewer: We're starting to standardize the error messages for enterprise features. In most cases, it will look like this:

```text
SAML: this feature requires Teleport Enterprise
```

expiration date of your license. 

Run the following command to inspect the expiration date of your license,
assuming your license is saved as `license.pem`:

```code
$ openssl x509 -text -in license.pem | grep "Not After"
             Not After : Dec 16 19:43:40 2022 GMT
```

### Renew or update your license

If you have changed your license agreement with Teleport, e.g., you have added
or removed support for a feature, you must obtain an updated license file and
replace your existing license file on all Teleport Auth Service instances. You
must also obtain a new license file after renewing your Teleport Enterprise
license.

Once you restart the Teleport Auth Service, any changes to the license will take
effect.

The Teleport Auth Service checks the expiration status of your license every
hour. This means that, after you replace your license file and restart your Auth
Service instances, it can take up to an hour to confirm the new license. 

## When your license expires

At 90 days before your Teleport Enterprise license expires, users will see an
alert similar to the following when running `tsh login`, `tsh status`, or any

> `tctl command`:

Reviewer 2: @ptgott can you clarifying what dismissing the message will do?

It should disable the message for the duration of the session.
